src/baldr/graphtile.cc

```diff
@@ -126,24 +129,51 @@ GraphTile::GraphTile(const GraphId& graphid, char* ptr, size_t size) : header_(n
   Initialize(graphid, ptr, size);
 }
 
-GraphTile::GraphTile(const std::string& tile_url, const GraphId& graphid, curler_t& curler) {
+GraphTile GraphTile::CacheTileURL(const std::string& tile_url,
+                                  const GraphId& graphid,
+                                  curler_t& curler,
+                                  bool gzipped,
+                                  const std::string& cache_location) {
   // Don't bother with invalid ids
   if (!graphid.Is_Valid() || graphid.level() > TileHierarchy::get_max_level()) {
-    return;
+    return {};
   }
 
-  // Get the response returned from curl
-  std::string uri =
-      tile_url + filesystem::path::preferred_separator + FileSuffix(graphid.Tile_Base());
+  // Get the response returned from curl, notice that when we expect zipped
+  // we tell curl not to unzip it, so that we can store it on the disk unzipped
+  auto suffix = FileSuffix(graphid.Tile_Base(), gzipped);
+  auto uri = tile_url + '/' + suffix;
   long http_code;
-  auto tile_data = curler(uri, http_code);
+  auto tile_data = curler(uri, http_code, false);
 
   // If its good try to use it
   if (http_code == 200) {
-    graphtile_ = std::make_shared<std::vector<char>>(std::move(tile_data));
-    Initialize(graphid, graphtile_->data(), graphtile_->size());
-    // TODO: optionally write the tile to disk?
+    // try to cache it on disk so we dont have to keep doing this
+    if (!cache_location.empty()) {
+      auto disk_location = cache_location + filesystem::path::preferred_separator + suffix;
+      auto dir = filesystem::path(cache_location + filesystem::path::preferred_separator + suffix);
+      dir.replace_filename("");
+      if (filesystem::create_directories(dir)) {
+        std::ofstream file(disk_location, std::ios::out | std::ios::binary | std::ios::ate);
+        file.write(&tile_data[0], tile_data.size());
+        return GraphTile(cache_location, graphid);
+      }
+    }
+
+    // and if we cant cache to disk try to use it in-memory
+    auto tile = GraphTile();
+    if (gzipped) {
+      tile.DecompressTile(graphid, tile_data);
+    } // we dont need to decompress so just take ownership of the data
+    else {
+      tile.graphtile_.reset(new std::vector<char>(0, 0));
```

Comment:
we also support non gzipped tiles, there is a new config parameter to tell which one to expect when curling. the reason you want this is so that you can cache a gzipped tile on disk isntead of always unpacking everything and storing uncompressed data on disk.